### PR TITLE
Detecting invalid user memory given as system call inputs

### PR DESCRIPTION
**Pull request by chiache**
This PR introduces two APIs, `test_user_memory()` and `test_user_string()`, to check if user inputs point to readable / writable virtual memory regions. If the tests failed, the system calls should return -EFAULT or -EINVAL according to the Linux specification.

### CODE REVIEW

File: LibOS/shim/src/bookkeep/shim_signal.c [hunk 1]
```diff
@@ -246,6 +246,15 @@ static void divzero_upcall (PAL_PTR event, PAL_NUM arg, PAL_CONTEXT * context)
 
 static void memfault_upcall (PAL_PTR event, PAL_NUM arg, PAL_CONTEXT * context)
```
Comment (donporter) [inline LibOS/shim/src/bookkeep/shim_signal.c]: Can you add some comments about the role of this function, when it is likely called, and how the arguments should be interpreted?

File: LibOS/shim/include/shim_tls.h [hunk 2]
```diff
@@ -73,6 +73,10 @@ typedef struct {
     void *                  debug_buf;
     int                     last_lock;
     struct lock_record      held_locks[NUM_LOCK_RECORD];
+    struct {
```
Comment (donporter) [inline LibOS/shim/include/shim_tls.h]: Can I get a comment on the purpose of this field?

File: LibOS/shim/src/bookkeep/shim_signal.c [hunk 3]
```diff
@@ -295,6 +304,71 @@ static void memfault_upcall (PAL_PTR event, PAL_NUM arg, PAL_CONTEXT * context)
     DkExceptionReturn(event);
 }
 
+bool test_user_memory (void * addr, size_t size, bool write)
```
Comment (donporter) [inline LibOS/shim/src/bookkeep/shim_signal.c]: Same request - add a comment documenting the function's job and expected behavior.

File: LibOS/shim/src/bookkeep/shim_signal.c [hunk 4]
```diff
@@ -295,6 +304,71 @@ static void memfault_upcall (PAL_PTR event, PAL_NUM arg, PAL_CONTEXT * context)
...
+
+    has_fault = false;
+
+ret_fault:
+    tcb->test_range.cont_addr = NULL;
+    tcb->test_range.start = tcb->test_range.end = NULL;
+    return has_fault;
+}
+
+bool test_user_string (const char * addr)
```
Comment (donporter) [inline LibOS/shim/src/bookkeep/shim_signal.c]: Ditto on function-level comment.

File: LibOS/shim/src/ipc/shim_ipc.c [hunk 5]
```diff
@@ -368,11 +368,20 @@ int send_ipc_message (struct shim_ipc_msg * msg, struct shim_ipc_port * port)
     assert(msg->size >= IPC_MSG_MINIMAL_SIZE);
     msg->src = cur_process.vmid;
 
-    int ret = DkStreamWrite(port->pal_handle, 0, msg->size, msg, NULL);
+    debug("send ipc message to port %p (handle %p)\n", port,
+          port->pal_handle);
+
+    PAL_HANDLE pal_handle = port->pal_handle;
+    rmb();
```
Comment (donporter) [inline LibOS/shim/src/ipc/shim_ipc.c]: Why is a barrier needed here?  This also merits a comment.

**Comment by donporter** [comment]
Also, I'm ok with removing lat_unix and the sendto LTP tests, and opening a separate issue to track that they need to eventually be put back.  I've found both to be a little flaky in CI.

**Comment by chiache** [comment]
I have added some comments to explain the intuition of this PR.

Basically, the `test_user_memory()` and `test_user_string()` functions are purely for compatibility, not security. In fact, I am not sure whether we need to / is able to provide any race-free memory checking against a malicious user. From the perspective of the library OS, we can only assume the users are benign and won't actively exploit the concurrency to attack the library OS. If the users are malicious, then the whole library OS is compromised. Something like `copy_from_user() `may stop unaware attackers but cannot stop more advanced attacks.

In short, you can assume that the main purpose of this PR is nothing but to revive a big chunk of LTP tests that are previously failing because they didn't get EFAULT as errno.

**Comment by donporter** [comment]
I am ok with the idea that the libOS cannot protect itself against an app that casts a pointer into libOS-internal memory and scribbles on it.  I suppose we might eventually hit an app that does a concurrent munmap or something.

I think this fix is an improvement, and we should take it.  It might be good to either open an issue documenting the concurrent munmap limitations, or maybe add a page on the wiki about this.

File: LibOS/shim/src/ipc/shim_ipc.c (hunk 5 above)
Comment (chiache) [reply LibOS/shim/src/ipc/shim_ipc.c]: I added some comment here, as well as in `shim_ipc_helper.c` where `port->deleted` is set to true. Basically we need the barrier here to ensure `port->deleted` is up-to-date in the memory.

File: LibOS/shim/src/bookkeep/shim_signal.c (hunk 1 above)
Comment (chiache) [reply LibOS/shim/src/bookkeep/shim_signal.c]: Already added.

File: LibOS/shim/include/shim_tls.h (hunk 2 above)
Comment (chiache) [reply LibOS/shim/include/shim_tls.h]: Added.

**Comment by donporter** [comment]
Test this please

**Comment by donporter** [comment]
The last CI job failed on dforkexec, which was disabled in master - not sure why Jenkins still ran this test.

**Comment by chiache** [comment]
@donporter I think the lmbench-2.5 directory still points to the previous commit.

**Comment by donporter** [comment]
I would have expected merging master to have picked up the latest LTP reference.
